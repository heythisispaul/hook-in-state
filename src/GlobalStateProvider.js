/* eslint-disable no-unused-vars */
import React, { createContext, useReducer } from 'react';
import {
  mapDispatchToReducers,
  getNextStateBuilder,
  getInStateBuilder,
  useSelectorBuilder,
  customDispatchBuilder,
} from './helpers'; 

export const GlobalStateContext = createContext();

const GlobalStateProvider = ({ children, reducers, initialState = {} }) => {
  const getNextState = getNextStateBuilder(reducers);
  const [state, dispatch] = useReducer(getNextState, initialState);
  const actions = mapDispatchToReducers(dispatch, reducers);
  const customDispatch = customDispatchBuilder(dispatch);
  const useSelector = useSelectorBuilder(state);
  const selectInState = getInStateBuilder(state);

  return (
    <GlobalStateContext.Provider
      value={{
        dispatch: customDispatch,
        actions,
        selectInState,
        useSelector,
      }}
    >
      {children}
    </GlobalStateContext.Provider>
  );
};

export default GlobalStateProvider;

