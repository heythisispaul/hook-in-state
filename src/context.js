/* eslint-disable no-unused-vars */
import React, { createContext } from 'react';

const GlobalState = createContext();

export const GlobalStateProvider = (contextValues) => ({ children }) => (
  <GlobalState.Provider value={contextValues}>
    {children}
  </GlobalState.Provider>
);

GlobalStateProvider.displayName = 'useGlobalState';

export default GlobalState;
