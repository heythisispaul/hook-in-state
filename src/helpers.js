import React from 'react';
import { produce } from 'immer';

export const getInStateBuilder = (state = {}) => (path, defaultValue) => {
  const parsedPath = Array.isArray(path) 
    ? path
    : path 
      .replace(/\[/g, '.')
      .replace(/]/g, '')
      .split('.')
      .filter(Boolean);
  let current = { ...state };
  const isValidPath = parsedPath.every((step) => {
    const stepIsValid = current[step] !== undefined;
    if (stepIsValid) {
      current = current[step];
    }
    return stepIsValid;
  });
  return isValidPath ? current : defaultValue;
};

export const mapDispatchToReducers = (dispatch, providedReducers = {}) => {
  const actions = {};
  Object.entries(providedReducers).forEach((currentAction) => {
    const [name] = currentAction;
    const type = name.toString();
    actions[type] = (value) => dispatch({ type, value });
  });
  return actions;
};

// const value =  useSelector((state) => state.count, []);
export const buildMemoCache = (memoArray, state) => memoArray.map((memoValue) => {
  return typeof memoValue === 'function' ? memoValue(state) : memoValue;
});

export const useSelectorBuilder = (state) => (selector, memoArray = []) => React.useMemo(
  () => selector(state), 
  buildMemoCache(memoArray, state)
);

export const customDispatchBuilder = (dispatch) => (value, reducer) => dispatch({ value, reducer });

export const getNextStateBuilder = (reducers = {}) => (state, action) => (
  produce(state, (draft) => {
    const { type, value, reducer } = action;
    const matchingReducer = reducers[type];
    if (matchingReducer) {
      matchingReducer(draft, value);
    } 
    if (reducer) {
      reducer(draft, value);
    }
  })
);