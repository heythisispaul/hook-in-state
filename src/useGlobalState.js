import React from 'react';
import { GlobalStateContext } from './GlobalStateProvider';

// not desctructuring for easier mocking
const useGlobalState = () => React.useContext(GlobalStateContext);

export default useGlobalState;
