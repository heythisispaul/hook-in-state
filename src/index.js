import useGlobalState from './useGlobalState';

export { default as GlobalStateProvider } from './GlobalStateProvider';
export default useGlobalState;
