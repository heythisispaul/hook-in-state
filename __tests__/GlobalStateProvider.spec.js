/* eslint-disable no-unused-vars */
import React from 'react';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import GlobalStateProvider from '../src/GlobalStateProvider';
import { createMockState } from './testHelpers/createMockState';
import { getInStateBuilder, mapDispatchToReducers } from '../src/helpers';

jest.mock('../src/helpers');

const dispatch = jest.fn();

configure({ adapter: new Adapter() });

beforeEach(() => {
  [getInStateBuilder, mapDispatchToReducers, dispatch]
    .forEach((mock) => mock.mockReset());
});

describe('GlobalStateProvider', () => {
  const ProviderRender = (reducers, state) => {
    const render = mount(
      <GlobalStateProvider reducers={reducers} initialState={state}>
        <div>I'll render!</div>
        <div>So will I!</div>
      </GlobalStateProvider>
    );
    return render;
  }
  const state = createMockState();
  const reducer = { test: (state) => state.count++ };
  it('provides the context providing component and it renders its children', () => {
    const render = ProviderRender(reducer, state);
    expect(render.html()).toMatchSnapshot();
  });

  it('should not error if no initialState is provided', () => {
    expect.assertions(2);
    const render = ProviderRender(reducer);
    expect(getInStateBuilder).toHaveBeenCalledWith({});
    expect(render.html()).toMatchSnapshot();
  });

  it('should map the provided reducers to dispatch', () => {
    ProviderRender(reducer);
    expect(mapDispatchToReducers.mock.calls[0][1]).toEqual(reducer);
  });
});
