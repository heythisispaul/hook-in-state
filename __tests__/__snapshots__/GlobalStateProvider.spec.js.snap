// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalStateProvider provides the context providing component and it renders its children 1`] = `"<div>I'll render!</div><div>So will I!</div>"`;

exports[`GlobalStateProvider should not error if no initialState is provided 1`] = `"<div>I'll render!</div><div>So will I!</div>"`;
